pkg/nsx/services/vpc/vpc.go

```diff
@@ -470,3 +521,232 @@ func (s *VPCService) Cleanup() error {
+			log.Info("avi rule not changed, skip update avi rule")
+			return nil
+		} else {
+			log.Info("avi rule changed", "previous", allowrule.DestinationGroups, "current", externalCIDR)
+		}
+	}
+
+	group, err := service.getorCreateAVIGroup(orgId, projectId, *vpc.Id, groupId)
+	if err != nil {
+		log.Error(err, "failed to get avi group", "group", groupId)
+		return err
+	}
+
+	newrule, err := service.buildAVIAllowRule(vpc, externalCIDR, *group.Path, ruleId, projectId)
+	log.Info("creating avi rule", "rule", newrule)
+	if err != nil {
+		log.Error(err, "failed to build avi rule", "rule", newrule)
+		return err
+	}
+
+	err = service.NSXClient.VPCRuleClient.Patch(orgId, projectId, *vpc.Id, spId, *newrule.Id, *newrule)
+	if err != nil {
+		log.Error(err, "failed to create avi rule", "rule", newrule)
+		return err
+	}
+	nsxrule, err := service.NSXClient.VPCRuleClient.Get(orgId, projectId, *vpc.Id, spId, *newrule.Id)
+	if err != nil {
+		log.Error(err, "failed to get avi rule", "rule", nsxrule)
+		return err
+	}
+	service.RuleStore.Add(nsxrule)
+	log.Info("created avi rule successfully")
+	return nil
+}
+
+func (service *VPCService) getorCreateAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	group, err := service.getAVIGroup(orgId, projectId, vpcId, groupId)
+	if err != nil {
+		log.Info("create avi group", "group", groupId)
+		group, err = service.createAVIGroup(orgId, projectId, vpcId, groupId)
+		if err != nil {
+			log.Error(err, "failed to create avi group", "group", groupId)
+			return group, err
+		}
+	}
+	return group, err
+}
+
+func (service *VPCService) buildAVIGroupTag(vpcId string) []model.Tag {
+	return []model.Tag{
+		{
+			Scope: common.String(common.TagScopeCluster),
+			Tag:   common.String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: common.String(common.TagScopeVersion),
+			Tag:   common.String(strings.Join(common.TagValueVersion, ".")),
+		},
+		{
+			Scope: common.String(common.TagScopeVPCCRUID),
+			Tag:   common.String(vpcId),
+		},
+		{
+			Scope: common.String(common.TagScopeGroupType),
+			Tag:   common.String(common.TagValueGroupAvi),
+		},
+	}
+}
+
+func (service *VPCService) createAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	group := model.Group{}
+	group.Tags = service.buildAVIGroupTag(vpcId)
+	expression := service.buildExpression("Condition", "VpcSubnet", "AVI_SUBNET_LB|", "Tag", "EQUALS", "EQUALS")
+	group.Expression = []*data.StructValue{expression}
+	group.DisplayName = common.String(groupId)
+
+	err := service.NSXClient.VpcGroupClient.Patch(orgId, projectId, vpcId, groupId, group)
+	if err != nil {
+		return &group, err
+	}
+	nsxgroup, err := service.NSXClient.VpcGroupClient.Get(orgId, projectId, vpcId, groupId)
+	return &nsxgroup, err
+}
+
+func (service *VPCService) buildExpression(resource_type, member_type, value, key, operator, scope_op string) *data.StructValue {
```

Comment:
there is buildExpression func in SecurityPolicyService, could we make it a common func?